### DISCUSSION TITLE
v3.0.0

### DISCUSSION THREAD

**Discussion by kennyadsl** [solidusio/solidus#4030] [Announcements]
🎉 Solidus 3.0 is out!

This version is almost identical to the last 2.11.x version with the only difference that we removed all the code deprecated. Here's a list of things to take into account while upgrading.

### Major Changes

**Removal of all code deprecated during the 2.x series**

The main change in this major version is the removal of all deprecated code that
we introduced during the 2.x series. This means that if any code that was deprecated
is still being used, the application will break. Following the deprecation messages
in the application logs, it should be quite easy to spot what code needs to be changed.

The main things that could break a Solidus application are:

***Paranoia gem has been replaced by Discard gem***

All references to methods added to models by Paranoia will raise a NoMethodError exception now.
Some of those methods are:

- `paranoia_destroy`
- `paranoia_delete`
- `with_deleted`
- `only_deleted`
- `really_destroy!`
- `after_real_destroy`

Pull Requests:

- Discard Paranoia [#3488](https://github.com/solidusio/solidus/pull/3488) ([cedum](https://github.com/cedum))

***Removed core support to first_name and last_name in Spree::Address***

In Solidus v2.11, we added a `name` attribute to `Spree::Address`, which is being populated combining
`first_name` and `last_name` values every time a new address is added to the system. We also provided
a rake task to update all existing records in order to get applications ready for Solidus 3.0.

With this major version, `name` is the only supported attributes. `first_name` and `last_name` fields are already in the database
so if needed, a store can revert this change by implementing its own logic.

See [3234](https://github.com/solidusio/solidus/issues/3234) for the rationale behind this change.

Pull Requests:

- Move Spree::Address#name attribute to the db [#3908](https://github.com/solidusio/solidus/pull/3908) ([filippoliverani](https://github.com/filippoliverani))
- Remove deprecated name-related Address fields [#3820](https://github.com/solidusio/solidus/pull/3820) ([filippoliverani](https://github.com/filippoliverani))

***All the other deprecations removal***

For a complete reference to rest of the code removed, these PRs can be taken as reference:

- Remove deprecated attachment_partial_name [#3974](https://github.com/solidusio/solidus/pull/3974) ([kennyadsl](https://github.com/kennyadsl))
- Remove legacy address state validation logic [#3847](https://github.com/solidusio/solidus/pull/3847) ([cedum](https://github.com/cedum))
- Raise canceling a payment when try_void is not implemented [#3844](https://github.com/solidusio/solidus/pull/3844) ([kennyadsl](https://
- Remove all code deprecated in Solidus v2.x [#3818](https://github.com/solidusio/solidus/pull/3818) ([kennyadsl](https://github.com/kennyadsl))

***Removal without deprecations***

We also removed some code that didn't need a deprecation warning. Be sure that your
codebase doesn't use any of the following:

- `Spree::LineItem::CurrencyMismatch` exception: we are not using it anymore since the behavior we had with `Spree::Config.raise_with_invalid_currency = true` has been removed.
- `Spree::Order::Checkout` is not used anymore. `Spree::Core::StateMachines::Order` is identical.
- `Spree::Admin::PaymentsHelper` module is empty after removing all deprecated methods inside it.
- `UserPaymentSource` is empty after removing all deprecated methods inside it.
- `Spree::Refund#perform_after_create` attribute, it was into a deprecated path. If you are still using it, please stop, it does nothing now.
- `Spree::TaxCalculator#ShippingRate`: it is always `nil` now.
- `Spree::Money::RUBY_NUMERIC_STRING`: was only used in a deprecated code path.

We also removed the following preferences without deprecations. They were just controlling a deprecated
flow and have no effect so, assuming you already switched to the only accepted value, you can safely
remove them from your initializer. You'll probably notice that because your app won't start.

- `Spree::Config.raise_with_invalid_currency`
- `Spree::Config.redirect_back_on_unauthorized preference`
- `Spree::Config.run_order_validations_on_order_updater preference `
- `Spree::Config.use_legacy_order_state_machine`
- `Spree::Config.use_legacy_store_credit_reimbursement_category_name`
- `Spree::Config.consider_actionless_promotion_active`
- `Spree::Config.use_legacy_address_state_validator`
- `Spree::Config.use_combined_first_and_last_name_in_address`

**By default, do not require the whole Rails framework**

This shouldn't give any issue in host applications, but if that happens,
it can be easily fixable opening `config/application.rb` and add `require 'rails/all'` or
the specific part of Rails needed by the application.

- Only require the necessary Rails frameworks [#3478](https://github.com/solidusio/solidus/pull/3478) ([elia](https://github.com/elia))

**Switch Paperclip dependency to its maintained version**

We recently added support for Active Support, which will be the default in Solidus 3.0.
Paperclip will still be around and supported for a while because we don't want to force
existing stores to accomplish the assets migration. While we support it, we want to use
the maintained fork.

- Switch to maintained Paperclip fork [#3913](https://github.com/solidusio/solidus/pull/3913) ([filippoliverani](https://github.com/filippoliverani))

### Misc

- Bump removal horizon for 3.x deprecations [#4025](https://github.com/solidusio/solidus/pull/4025) ([kennyadsl](https://github.com/kennyadsl))
- Add Post-Install message to Solidus 3.0 [#3985](https://github.com/solidusio/solidus/pull/3985) ([kennyadsl](https://github.com/kennyadsl))
- Add Active Storage in Dummy App for extensions [#3969](https://github.com/solidusio/solidus/pull/3969) ([kennyadsl](https://github.com/kennyadsl))
- Improve Active Storage configuration for in-memory Dummy App [#3970](https://github.com/solidusio/solidus/pull/3970) ([kennyadsl](https://github.com/kennyadsl))
- Allow users to create blank issues in GitHub [#3943](https://github.com/solidusio/solidus/pull/3943) ([kennyadsl](https://github.com/kennyadsl))
- Install Active Storage by default on new stores [#3938](https://github.com/solidusio/solidus/pull/3938) ([kennyadsl](https://github.com/kennyadsl))
- Avoid too many prompts during solidus:install generator [#3937](https://github.com/solidusio/solidus/pull/3937) ([kennyadsl](https://github.com/kennyadsl))
- Align Rubocop ruby version to gemspec [#3935](https://github.com/solidusio/solidus/pull/3935) ([spaghetticode](https://github.com/spaghetticode))
- Skip adding webpacker gem when generating dummyapp [#3922](https://github.com/solidusio/solidus/pull/3922) ([SamuelMartini](https://github.com/SamuelMartini))
- allow customize database credentials for test app [#3921](https://github.com/solidusio/solidus/pull/3921) ([ccarruitero](https://github.com/ccarruitero))
- Bump redcarpet from 3.4.0 to 3.5.1 in /guides [#3890](https://github.com/solidusio/solidus/pull/3890) ([dependabot](https://github.com/apps/dependabot))
- Adjust CircleCI config to reflect Rails versions that we support [#3885](https://github.com/solidusio/solidus/pull/3885) ([kennyadsl](https://github.com/kennyadsl))
- Bump axios from 0.18.1 to 0.21.1 in /guides [#3881](https://github.com/solidusio/solidus/pull/3881) ([dependabot](https://github.com/apps/dependabot))
- Bump ini from 1.3.5 to 1.3.7 in /guides [#3861](https://github.com/solidusio/solidus/pull/3861) ([dependabot](https://github.com/apps/dependabot))
- Drive community to GitHub Discussions when opening issues [#3857](https://github.com/solidusio/solidus/pull/3857) ([kennyadsl](https://github.com/kennyadsl))
github.com/kennyadsl))
- Update governance with latest changes to the organization [#3836](https://github.com/solidusio/solidus/pull/3836) ([kennyadsl](https://github.com/kennyadsl))
- Fix install instructions in Solidus Guides [#3833](https://github.com/solidusio/solidus/pull/3833) ([ikraamg](https://github.com/ikraamg))
- Update install instructions after 2.11 release [#3825](https://github.com/solidusio/solidus/pull/3825) ([kennyadsl](https://github.com/kennyadsl))
- Move "thinking cat" fixture to lib folder [#3824](https://github.com/solidusio/solidus/pull/3824) ([mamhoff](https://github.com/mamhoff))
- Update readme with Solidus demo URL [#3822](https://github.com/solidusio/solidus/pull/3822) ([seand7565](https://github.com/seand7565))
- Fix headers in changelog [#3812](https://github.com/solidusio/solidus/pull/3812) ([jarednorman](https://github.com/jarednorman))
- Fixed typo with misspell [#3811](https://github.com/solidusio/solidus/pull/3811) ([hsbt](https://github.com/hsbt))

<hr /><em>This discussion was created from the release <a href='https://github.com/solidusio/solidus/releases/tag/v3.0.0'>v3.0.0</a>.</em>

## Suggested Upgrade Path

1. Update to the last released version of Solidus `v2.11` (as of today, it is `2.11.9`).
2. Be sure you have run the migrations installer with `bin/rails railties:install:migrations` and `bin/rails db:migrate`.
3. Run the update task with `bin/rails solidus:upgrade:two_point_eleven`.
4. Check your logs (either running your test suite or your production logs) and remove any deprecation warning. If you are using deprecated code, it is very likely that your application is not compliant with Solidus 3.0. 
5. Update your Gemfile to use `gem 'solidus', '~> 3.0.0'`, run `bundle install`.

## Common problems

👇 I'll leave some common errors in the comments below with the solution. Feel free to comment here if you encounter any problem upgrading.

**Comment by kennyadsl** [comment]
#### Spree::Image customization

**Error:**

```text
~/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/activerecord-6.1.3/lib/active_record/dynamic_matchers.rb:22:in `method_missing': undefined method `has_one_attached' for #<Class:0x00007fc91f548d38>
Did you mean?  has_attached_file (NoMethodError)
```

**Cause**

`Spree::Image` has now its default adapter to Active Storage, and you probably want to keep using the Paperclip adapter. This setting can be changed with: 

```ruby
# config/initializers/spree.rb`

config.image_attachment_module = 'Spree::Image::PaperclipAttachment'
config.taxon_attachment_module = 'Spree::Taxon::PaperclipAttachment'
```

Take care you don't have any customization of `Spree::Image.attachment_definitions` before that preference (even in other initializer files loaded before, eg `config/initializers/paperclip.rb`). If you do, you'll see the error described above. 

**Solution**

To fix, move the customization to a decorator, for example, if you have:

```ruby
# config/intitializers/something.rb

 Spree::Image.attachment_definitions[:attachment][:styles] = {
   mini: '48x48>',
   small: '100x100>',
   product: '240x240>',
   large: '600x600>'
 }
```

Move it into:

```ruby
# app/models/spree/image_decorator.rb

# frozen_string_literal: true
module Spree
  module ImageDecorator
    class << self
      private

      def prepended(klass)
        klass.attachment_definitions[:attachment][:styles] = {
          mini: '48x48>',
          small: '100x100>',
          product: '240x240>',
          large: '600x600>'
        }
      end
    end

    Spree::Image.prepend self
  end
end
```

**Reply by afdev82** [reply]
I was following [this page](https://guides.solidus.io/developers/products-and-variants/product-images.html) instructions to regenerate the thumbnails (I was waiting to update Solidus to v3 to use the forked Paperclip due to [this bug](https://github.com/thoughtbot/paperclip/pull/2637)).
But the task is not working, I followed your instructions to move my configuration in a decorator and specify the adapter, but the task does nothing.

This was my paperclip initializer before:

```
# In theory you must regenerate the paperclip thumbnails by running this command:
# bundle exec rake paperclip:refresh:thumbnails CLASS=Spree::Image
# BUT THIS IS NOT GOING TO WORK WITH AMAZON S3, see https://github.com/thoughtbot/paperclip/pull/2637 we need to wait Solidus v3 to support the fork of paperclip with the bug fix

Spree::Image.attachment_definitions[:attachment][:styles] = {
  mini: '167x118>',
  small: '213x165>',
  product: '235x235>',
  large: '1920x1080>'
}

if Rails.env.production? || Rails.env.staging?
  Paperclip::Attachment.default_options.merge!(
    storage: :fog,
    fog_credentials: {
      provider: 'AWS',
      aws_access_key_id: ENV['AWS_ACCESS_KEY_ID'],
      aws_secret_access_key: ENV['AWS_SECRET_ACCESS_KEY'],
      region: ENV['AWS_REGION']
    },
    fog_directory: ENV['AWS_S3_SOLIDUS_BUCKET_NAME'],
    path: "#{Rails.env}/solidus/products/:id/:style/:basename.:extension",
    preserve_files: true
  )

  Spree::Image.attachment_definitions[:attachment].delete(:url)
  Spree::Image.attachment_definitions[:attachment].delete(:path)
end
```

and after:

```
if Rails.env.production? || Rails.env.staging?
  Paperclip::Attachment.default_options.merge!(
    storage: :fog,
    fog_credentials: {
      provider: 'AWS',
      aws_access_key_id: ENV['AWS_ACCESS_KEY_ID'],
      aws_secret_access_key: ENV['AWS_SECRET_ACCESS_KEY'],
      region: ENV['AWS_REGION']
    },
    fog_directory: ENV['AWS_S3_SOLIDUS_BUCKET_NAME'],
    path: "#{Rails.env}/solidus/products/:id/:style/:basename.:extension",
    preserve_files: true
  )
end
```

plus the decorator:
```
module Spree
  module ImageDecorator
    class << self
      private

      def prepended(klass)
        # If changed, you must regenerate the paperclip thumbnails by running this command:
        # bundle exec rake paperclip:refresh:thumbnails CLASS=Spree::Image
        # This only works with the fork of the original Paperclip, see https://github.com/thoughtbot/paperclip/pull/2637
        klass.attachment_definitions[:attachment][:styles] = {
          mini: '167x118>',
          small: '213x165>',
          product: '235x235>',
          large: '1920x1080>'
        }

        if Rails.env.production? || Rails.env.staging?
          klass.attachment_definitions[:attachment].delete(:url)
          klass.attachment_definitions[:attachment].delete(:path)
        end
      end
    end

    Spree::Image.prepend self
  end
end
```

If I inspect the styles in the console I get the new ones:

```
irb(main):001:0> Spree::Image.attachment_definitions[:attachment][:styles]
=> {:mini=>"167x118>", :small=>"213x165>", :product=>"235x235>", :large=>"1920x1080>"}
```

**Reply by kennyadsl** [reply]
What do you mean by "the task is not working"? Any specific error message to share?

**Reply by afdev82** [reply]
It was just doing nothing. The task is working now.
I had also problems (an error about missing asset for the Asset pipeline) setting up the CDN using the :fog storage option for paperclip, so I switched to :s3 and it's working now. I don't think it was related to the Solidus update.
Sorry for the noise.

**Comment by kennyadsl** [comment]
### ActionCable JS inclusion

**Error**

```
ActionView::Template::Error: couldn't find file 'action_cable' with type 'application/javascript'
```

**Cause**

We are not requiring the whole `rails/all` bundle anymore in Solidus. See https://github.com/solidusio/solidus/pull/3478.

**Solution**

Depending on if you are using ActionCable or not, you may want to manually `require 'action_cable'` in your `config/application.rb` or just remove `//= require action_cable` somewhere in your assets pipeline manifests.

**Comment by alexisagos** [comment]
Hi, I just upgraded from **3.0.0.rc2** to **3.0.0** and now when I start my server I'm getting the following logs (fresh install, no specific graphql changes):

```
Booting Puma
=> Rails 6.1.3.1 application starting in development 
=> Run `bin/rails server --help` for more startup options
Puma starting in single mode...
* Puma version: 5.2.2 (ruby 2.7.2-p137) ("Fettisdagsbulle")
*  Min threads: 5
*  Max threads: 5
*  Environment: development
*          PID: 16766
* Listening on http://127.0.0.1:3000
* Listening on http://[::1]:3000
Use Ctrl-C to stop
Started GET "/graphql" for ::1 at 2021-04-29 21:33:19 +0200
   (3.6ms)  SELECT sqlite_version(*)
   (0.3ms)  SELECT "schema_migrations"."version" FROM "schema_migrations" ORDER BY "schema_migrations"."version" ASC
  
ActionController::RoutingError (No route matches [GET] "/graphql"):
  
Started GET "/graphql" for ::1 at 2021-04-29 21:33:21 +0200
  
ActionController::RoutingError (No route matches [GET] "/graphql"):
  
Started GET "/graphql" for ::1 at 2021-04-29 21:33:25 +0200
  
ActionController::RoutingError (No route matches [GET] "/graphql"):
  
Started GET "/graphql" for ::1 at 2021-04-29 21:33:27 +0200
  
ActionController::RoutingError (No route matches [GET] "/graphql"):
  
Started GET "/graphql" for ::1 at 2021-04-29 21:33:30 +0200
  
ActionController::RoutingError (No route matches [GET] "/graphql"):
  
Started GET "/graphql" for 127.0.0.1 at 2021-04-29 21:33:31 +0200
  
ActionController::RoutingError (No route matches [GET] "/graphql"):
  
Started GET "/graphql" for ::1 at 2021-04-29 21:33:33 +0200
  
ActionController::RoutingError (No route matches [GET] "/graphql"):
  
Started GET "/graphql" for ::1 at 2021-04-29 21:33:37 +0200
  
ActionController::RoutingError (No route matches [GET] "/graphql"):
  
Started GET "/graphql" for ::1 at 2021-04-29 21:33:39 +0200
  
```

Apart from these weird logs, app is running and working well, does anyone have an idea what is causing this ?

Thanks !

**Reply by kennyadsl** [reply]
Thanks for reporting! This doesn't seem related to Solidus itself, are you using the [`solidus_graphql_api`](https://github.com/solidusio-contrib/solidus_graphql_api) extension? Looks like something is calling `/graphql`, but I can't think of anything that does that in Solidus or in that extension. Maybe it's something with your custom JS?

**Reply by alexisagos** [reply]
You're right I must have had another process running elsewhere causing this because these requests don't show any more today. Thanks for your quick answer !
